### PR TITLE
Better receiver types for static dispatch.

### PR DESCRIPTION
**Pull request by sjrd**
~~Based on #116. Only the last 3 commits belong to this PR.~~

### CODE REVIEW

**Comment by sjrd** [comment]
Rebased.

**Comment by sjrd** [comment]
Ping @tanishiking about this one?

**Comment by tanishiking** [comment]
Oops, I'll take a look today. thanks for pinging

File: wasm/src/main/scala/ir2wasm/WasmBuilder.scala [hunk 1]
```diff
@@ -1111,7 +1114,48 @@ class WasmBuilder(coreSpec: CoreSpec) {
     val body = method.body.getOrElse(throw new Exception("abstract method cannot be transformed"))
     WasmExpressionBuilder.generateIRBody(body, method.resultType)
 
-    fctx.buildAndAddToContext(useFunctionTypeInMainRecType = true)
+    fctx.buildAndAddToContext()
+
+    if (namespace == IRTrees.MemberNamespace.Public && !isHijackedClass) {
+      /* Also generate the bridge that is stored in the table entries. In table
```
Comment (tanishiking) [inline wasm/src/main/scala/ir2wasm/WasmBuilder.scala:1120]: :+1:

**Review by tanishiking — APPROVED** [review]
LGTM, thanks!

**Pull request merged**